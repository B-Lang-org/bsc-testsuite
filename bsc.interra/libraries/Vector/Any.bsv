package Any;

import Vector :: *;
function Bool f (Int #(32) a);
     if (a < 0)
        return True;
     else
        return False;
endfunction


module mkTestbench_Any();
   Vector #(5,Int #(32)) my_list1 = cons (0, cons (-1, cons (2, cons (-3, cons (4, nil)))));
   Vector #(3,Int #(32)) my_list2 = cons (0, cons (2, cons (4, nil)));



   rule fire_once (True);
      if (any(f, my_list1) != True || any(f, my_list2) != False) 
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
	  $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_Any
endpackage : Any
