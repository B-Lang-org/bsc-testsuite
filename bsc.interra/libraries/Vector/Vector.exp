test_c_veri_bsv_multi All mkTestbench_All {} "mkTestbench_All.out.expected"

test_c_veri_bsv_multi Any mkTestbench_Any {} "mkTestbench_Any.out.expected"

test_c_veri_bsv_multi Append mkTestbench_Append {} "mkTestbench_Append.out.expected"

test_c_veri_bsv_multi Concat mkTestbench_Concat {} "mkTestbench_Concat.out.expected" 

test_c_veri_bsv_multi Elem mkTestbench_Elem {} "mkTestbench_Elem.out.expected" 

test_c_veri_bsv_multi Equality mkTestbench_Equality {} "mkTestbench_Equality.out.expected" 

test_c_veri_bsv_multi FoldOrder mkTestbench_FoldOrder {} "mkTestbench_FoldOrder.out.expected" 

test_c_veri_bsv_multi Fold mkTestbench_Fold {} "mkTestbench_Fold.out.expected" 

test_c_veri_bsv_multi FoldM mkTestbench_FoldM {} "mkTestbench_FoldM.out.expected" 

test_c_veri_bsv_multi Foldl mkTestbench_Foldl {} "mkTestbench_Foldl.out.expected" 

test_c_veri_bsv_multi Foldl1 mkTestbench_Foldl1 {} "mkTestbench_Foldl1.out.expected" 

test_c_veri_bsv_multi FoldlM mkTestbench_FoldlM {} "mkTestbench_FoldlM.out.expected" 

test_c_veri_bsv_multi Foldr mkTestbench_Foldr {} "mkTestbench_Foldr.out.expected" 

test_c_veri_bsv_multi Foldr1 mkTestbench_Foldr1 {} "mkTestbench_Foldr1.out.expected" 

test_c_veri_bsv_multi FoldrM mkTestbench_FoldrM {} "mkTestbench_FoldrM.out.expected" 

test_c_veri_bsv_multi GenList mkTestbench_GenList {} "mkTestbench_GenList.out.expected" 

test_c_veri_bsv_multi GenWith mkTestbench_GenWith {} "mkTestbench_GenWith.out.expected" 

test_c_veri_bsv_multi GenWithM mkTestbench_GenWithM {} "mkTestbench_GenWithM.out.expected" 

test_c_veri_bsv_multi Head mkTestbench_Head {} "mkTestbench_Head.out.expected" 

test_c_veri_bsv_multi Init mkTestbench_Init {} "mkTestbench_Init.out.expected" 

test_c_veri_bsv_multi Last mkTestbench_Last {} "mkTestbench_Last.out.expected" 

test_c_veri_bsv_multi Map mkTestbench_Map {} "mkTestbench_Map.out.expected" 

test_c_veri_bsv_multi MapAccumL mkTestbench_MapAccumL {} "mkTestbench_MapAccumL.out.expected" 

test_c_veri_bsv_multi MapAccumR mkTestbench_MapAccumR {} "mkTestbench_MapAccumR.out.expected" 

test_c_veri_bsv_multi MapM mkTestbench_MapM {} "mkTestbench_MapM.out.expected" 

test_c_veri_bsv_multi MapM_ mkTestbench_MapM_ {} "mkTestbench_MapM_.out.expected" 

test_c_veri_bsv_multi MapPairs mkTestbench_MapPairs {} "mkTestbench_MapPairs.out.expected" 

test_c_veri_bsv_multi Replicate mkTestbench_Replicate {} "mkTestbench_Replicate.out.expected" 

test_c_veri_bsv_multi ReplicateM mkTestbench_ReplicateM {} "mkTestbench_ReplicateM.out.expected" 

test_c_veri_bsv_multi Reverse mkTestbench_Reverse {} "mkTestbench_Reverse.out.expected" 

test_c_veri_bsv_multi Rotate mkTestbench_Rotate {} "mkTestbench_Rotate.out.expected" 

test_c_veri_bsv_multi RotateR mkTestbench_RotateR {} "mkTestbench_RotateR.out.expected" 

test_c_veri_bsv_multi Scanl mkTestbench_Scanl {} "mkTestbench_Scanl.out.expected" 

test_c_veri_bsv_multi Scanr mkTestbench_Scanr {} "mkTestbench_Scanr.out.expected" 

test_c_veri_bsv_multi Select mkTestbench_Select {} "mkTestbench_Select.out.expected" 

if {$ctest == 1} {
compile_object_fail Sequence.bsv mkTestbench_Sequence
compare_file Sequence.bsv.bsc-ccomp-out
}
#link_objects_pass {"mkTestbench_Sequence"} mkTestbench_Sequence
#sim_final_state mkTestbench_Sequence 10010
#compare_file "mkTestbench_Sequence.final-state"

test_c_veri_bsv_multi Sscanl mkTestbench_Sscanl {} "mkTestbench_Sscanl.out.expected" 

test_c_veri_bsv_multi Sscanr mkTestbench_Sscanr {} "mkTestbench_Sscanr.out.expected" 

test_c_veri_bsv_multi Tail mkTestbench_Tail {} "mkTestbench_Tail.out.expected" 

test_c_veri_bsv_multi Take mkTestbench_Take {} "mkTestbench_Take.out.expected" 

test_c_veri_bsv_multi ToList mkTestbench_ToList {} "mkTestbench_ToList.out.expected" 

test_c_veri_bsv_multi ToListN mkTestbench_ToListN {} "mkTestbench_ToListN.out.expected" 

test_c_veri_bsv_multi Transpose mkTestbench_Transpose {} "mkTestbench_Transpose.out.expected" 

test_c_veri_bsv_multi TransposeLN mkTestbench_TransposeLN {} "mkTestbench_TransposeLN.out.expected" 

test_c_veri_bsv_multi UnZip mkTestbench_UnZip {} "mkTestbench_UnZip.out.expected" 

test_c_veri_bsv_multi Update mkTestbench_Update {} "mkTestbench_Update.out.expected" 

test_c_veri_bsv_multi Zip mkTestbench_Zip {} "mkTestbench_Zip.out.expected" 

test_c_veri_bsv_multi Zip3 mkTestbench_Zip3 {} "mkTestbench_Zip3.out.expected" 

test_c_veri_bsv_multi Zip4 mkTestbench_Zip4 {} "mkTestbench_Zip4.out.expected" 

test_c_veri_bsv_multi ZipAny mkTestbench_ZipAny {} "mkTestbench_ZipAny.out.expected" 

test_c_veri_bsv_multi ZipWith mkTestbench_ZipWith {} "mkTestbench_ZipWith.out.expected" 

test_c_veri_bsv_multi ZipWith3 mkTestbench_ZipWith3 {} "mkTestbench_ZipWith3.out.expected" 

test_c_veri_bsv_multi ZipWith3M mkTestbench_ZipWith3M {} "mkTestbench_ZipWith3M.out.expected" 

test_c_veri_bsv_multi ZipWithAny mkTestbench_ZipWithAny {} "mkTestbench_ZipWithAny.out.expected" 

test_c_veri_bsv_multi ZipWithAny3 mkTestbench_ZipWithAny3 {} "mkTestbench_ZipWithAny3.out.expected" 

test_c_veri_bsv_multi ZipWithM mkTestbench_ZipWithM {} "mkTestbench_ZipWithM.out.expected" 

test_c_veri_bsv_multi ZipWithM_ mkTestbench_ZipWithM_ {} "mkTestbench_ZipWithM_.out.expected" 

