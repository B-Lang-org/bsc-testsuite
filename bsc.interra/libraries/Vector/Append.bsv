package Append;

import Vector :: *;

function Action displayabc (a abc) provisos (Bits #(a, sa));
    action
      $display ("%d", abc);
    endaction
endfunction



function Action display_list (Vector #(n,a) my_list) provisos (Bits # (a,sa));
     action
       joinActions ( map (displayabc, my_list));
     endaction
endfunction



module mkTestbench_Append();
   Vector #(5,Int #(5)) my_list1 = cons (0, cons (1, cons (2, cons (3, cons (4, nil)))));
   Vector #(5,Int #(5)) my_list2 = cons (5, cons (6, cons (7, cons (8, cons (9, nil)))));
   Vector #(10,Int #(5)) my_list3 = cons (0, cons (1, cons (2, cons (3, cons (4, cons (5, cons (6, cons (7, cons (8, cons (9, nil))))))))));


   rule fire_once (True);
      $display("ListN1:");
      display_list (my_list1);
      $display("ListN2:");
      display_list (my_list2);
      $display("Appended Vector:");
      display_list (append(my_list1, my_list2));
      if (my_list3 != append (my_list1, my_list2)) 
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
	  $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_Append
endpackage : Append
