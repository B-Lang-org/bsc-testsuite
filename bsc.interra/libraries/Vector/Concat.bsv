package Concat;

import Vector :: *;

function Action displayabc (a abc) provisos (Bits #(a, sa));
    action
      $display ("%d", abc);
    endaction
endfunction



function Action display_list (Vector #(n,a) my_list) provisos (Bits # (a,sa));
     action
       joinActions ( map (displayabc, my_list));
     endaction
endfunction



module mkTestbench_Concat();
   Vector #(5,Int #(32)) my_list1 = cons (0, cons (1, cons (2, cons (3, cons (4, nil)))));
   Vector #(5,Int #(32)) my_list2 = cons (5, cons (6, cons (7, cons (8, cons (9, nil)))));
   Vector #(5,Int #(32)) my_list3 = cons (10, cons (11, cons (12, cons (13, cons (14, nil)))));
   Vector #(5,Int #(32)) my_list4 = cons (15, cons (16, cons (17, cons (18, cons (19, nil)))));

   Vector #(4,Vector #(5,Int #(32))) my_list5 = cons (my_list1, cons (my_list2, cons (my_list3, cons (my_list4, nil))));

   Vector #(20,Int #(32)) my_list6 = cons(0, cons (1, cons (2, cons (3, cons (4, cons (5, cons (6, cons (7, cons (8, cons (9, cons (10, cons (11, cons (12, cons (13, cons(14, cons (15, cons (16, cons (17, cons (18, cons (19, nil))))))))))))))))))));


   rule fire_once (True);
      $display ("Concatenated Vector:");
      display_list (concat(my_list5));
      if (concat(my_list5) != my_list6) 
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
	  $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_Concat
endpackage : Concat
