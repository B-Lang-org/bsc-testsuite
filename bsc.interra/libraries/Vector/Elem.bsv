package Elem;

import Vector :: *;

module mkTestbench_Elem();
   Vector #(5,Int #(32)) my_list1 = cons (0, cons (-1, cons (2, cons (3, cons (4, nil)))));


   rule fire_once (True);
      if (elem(0, my_list1) !=True ||  elem(1, my_list1) !=False || elem(2, my_list1) !=True || elem(3, my_list1) !=True || elem(5, my_list1) !=False || elem(4, my_list1) !=True || elem(-1, my_list1) !=True || elem(0, my_list1) !=True || elem(6, my_list1) !=False)
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
	  $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_Elem
endpackage : Elem
