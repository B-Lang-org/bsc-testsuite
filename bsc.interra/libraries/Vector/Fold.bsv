package Fold;

import Vector :: *;

function Action displayabc (a abc) provisos (Bits #(a, sa));
    action
      $display ("%d", abc);
    endaction
endfunction



function Action display_list (Vector #(n,a) my_list) provisos (Bits # (a,sa));
     action
       joinActions ( map (displayabc, my_list));
     endaction
endfunction

function Int #(5) add (Int #(5) a, Int #(5) b);
    return (a + b);
endfunction

module mkTestbench_Fold();
   Vector #(5,Int #(5)) my_list1 = cons (1, cons (2, cons (3, cons (4, cons (5, nil)))));


  
   rule fire_once (True);
      $display("Sum of list of first five integers = %d", fold (add, my_list1));
      if (fold(add,my_list1) != 15) 
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
	  $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_Fold
endpackage : Fold
