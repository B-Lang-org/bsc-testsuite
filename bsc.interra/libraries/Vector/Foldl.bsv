package Foldl;

import Vector :: *;

function Action displayabc (a abc) provisos (Bits #(a, sa));
    action
      $display ("%d", abc);
    endaction
endfunction



function Action display_list (Vector #(n,a) my_list) provisos (Bits # (a,sa));
     action
       joinActions ( map (displayabc, my_list));
     endaction
endfunction

function Int #(5) add (Int #(5) b, Integer a);
    return (fromInteger (a) + b);
endfunction

module mkTestbench_Foldl();
   Vector #(5,Integer) my_list1 = cons (1, cons (2, cons (3, cons (4, cons (5, nil)))));


  
   rule fire_once (True);
      $display("Sum of list of first five integers = %d", foldl (add, 0, my_list1));
      if (foldl(add, 0, my_list1) != 15) 
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
	  $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_Foldl
endpackage : Foldl
