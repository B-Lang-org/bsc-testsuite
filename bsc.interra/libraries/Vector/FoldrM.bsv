package FoldrM;

import Vector :: *;

function Action displayabc (a abc) provisos (Bits #(a, sa));
    action
      $display ("%d", abc);
    endaction
endfunction



function Action display_list (Vector #(n,a) my_list) provisos (Bits # (a,sa));
     action
       joinActions ( map (displayabc, my_list));
     endaction
endfunction

function ActionValue#(Int #(32)) add (Int #(32) a, Int #(32) b);
	actionvalue
	  noAction;
      return (a + b);
	endactionvalue
endfunction

module mkTestbench_FoldrM();
   Vector #(5,Int #(32)) my_list1 = cons (1, cons (2, cons (3, cons (4, cons (5, nil)))));



  
   rule fire_once (True);
      Int #(32) value <- foldrM(add,5,my_list1);
      $display("%d", value);
      if (value != 20) 
        $display ("Simulation Fails");
      else
        $display ("Simulation Passes");
	  $finish(2'b00);
   endrule 
      
endmodule : mkTestbench_FoldrM
endpackage : FoldrM
