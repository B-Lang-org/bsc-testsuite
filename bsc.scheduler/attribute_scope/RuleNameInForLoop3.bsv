import Vector::*;

(* synthesize *)
module sysRuleNameInForLoop3();
   Vector#(4, Reg#(UInt#(16))) rg <- replicateM(mkRegU);

   for (Integer i = 0; i < 4; i = i + 1) begin
      rule r1;
         $display("r1");
      endrule

      rule r2;
         rg[i] <= rg[i] + 1;
         $display("r2");
      endrule

      // By putting a non-rule statement here, we break them into two blocks
      Bool b = True;

      rule r3;
         $display("r3");
      endrule

      // Because there is no "r2" in this block, the attribute will refer
      // to a rule by that name already added to the module.  That means
      // that later iterations will refer to "r2" from the first iteration,
      // not the "r2" from the current iteration.
      //
      (* descending_urgency = "r2, r4" *)
      rule r4;
         rg[i] <= rg[i] + 2;
         $display("r4");
      endrule
   end
endmodule
