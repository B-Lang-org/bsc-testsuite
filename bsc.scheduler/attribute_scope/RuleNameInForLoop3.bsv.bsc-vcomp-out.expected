checking package dependencies
compiling RuleNameInForLoop3.bsv
code generation for sysRuleNameInForLoop3 starts
Warning: "RuleNameInForLoop3.bsv", line 4, column 8: (G0010)
  Rule "r2_1" was treated as more urgent than "r4_1". Conflicts:
    "r2_1" cannot fire before "r4_1": calls to rg_1.write vs. rg_1.read
    "r4_1" cannot fire before "r2_1": calls to rg_1.write vs. rg_1.read
Warning: "RuleNameInForLoop3.bsv", line 4, column 8: (G0010)
  Rule "r2_2" was treated as more urgent than "r4_2". Conflicts:
    "r2_2" cannot fire before "r4_2": calls to rg_2.write vs. rg_2.read
    "r4_2" cannot fire before "r2_2": calls to rg_2.write vs. rg_2.read
Warning: "RuleNameInForLoop3.bsv", line 4, column 8: (G0010)
  Rule "r2_3" was treated as more urgent than "r4_3". Conflicts:
    "r2_3" cannot fire before "r4_3": calls to rg_3.write vs. rg_3.read
    "r4_3" cannot fire before "r2_3": calls to rg_3.write vs. rg_3.read
Warning: "RuleNameInForLoop3.bsv", line 30, column 12: (G0021)
  According to the generated schedule, rule `r4' can never fire.
Warning: "RuleNameInForLoop3.bsv", line 30, column 12: (G0021)
  According to the generated schedule, rule `r4_1' can never fire.
Warning: "RuleNameInForLoop3.bsv", line 30, column 12: (G0021)
  According to the generated schedule, rule `r4_2' can never fire.
Warning: "RuleNameInForLoop3.bsv", line 30, column 12: (G0021)
  According to the generated schedule, rule `r4_3' can never fire.
Verilog file created: sysRuleNameInForLoop3.v
All packages are up to date.
